docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -201,44 +201,117 @@ The ShootState synchronization controller will become part of the gardenlet. It
+* `gardenlet` cannot connect to the seed's `kube-apiserver`. In this case `gardenlet` knows if migration has been triggered, but it will not start shoot migration or reconciliation as it will first check the seed conditions and try to update the `Cluster` resource, both of which will fail. Extension controllers could still be able to connect to the seed's `kube-apiserver` (if they are not running where `gardenlet` is running), and similarly to the previous case, they could still attempt to reconcile their resources.
+* The seed components (`etcd-druid`, extension controllers, etc) cannot connect to the seed's `kube-apiserver`. In this case extension controllers would not be able to reconcile their resources as they cannot fetch them from the seed's `kube-apiserver`. When the connection to the `kube-apiserver` comes back, the controllers might be stuck in a retry loop from a previous reconciliation, or the resources could still be annotated with `gardener.cloud/operation=reconcile`. This could lead to a race condition depending on who manages to `update` or `get` the resources first. If `gardenlet` manages to update the resources before they are read by the extension controllers, they would be properly updated with `gardener.cloud/operation=migrate`. Otherwise, they would be reconciled as usual.
+
+The safety mechanism is based on "cluster leases". Essentially, extension controllers must check if the seed has a valid "lease" for the shoot before attempting a reconciliation. The lease is just an additional timestamp field in the `Cluster` resource:
+
+```go
+// ClusterSpec is the spec for a Cluster resource.
+type ClusterSpec struct {
+	...
+	// LeaseExpiration indicates when the lease of this cluster will expire
+	LeaseExpiration metav1.MicroTime `json:"leaseExpiration"`
+}
+```
 
-When case 3 happens it means that the shoot has been migrated and the backuprestore container in the  __Destination Seed__ has started using the Shoot's backup entry. The backuprestore on the __Source Seed__ should be configured to shut itself and the etcd container down once it sees the newer timestamp. Shutting down the etcd will cause the kubernetes control plane components to go into crashloop backoff and the MCM will not be able to do anything as it will not be able to list nodes (this has to be verified with MCM).
+On a regular basis, e.g. every 2 minutes, and at each shoot reconciliation, `gardenlet` fetches the shoot from the Garden `kube-apiserver`, and if successful and the control plane migration hasn't been triggered it updates the `Cluster` resource with a renewed timestamp, e.g. 2 minutes from now. The extension controllers fetch this resource before each reconciliation and check if the lease has expired. If it hasn't, they reconcile as usual. If the lease has expired, they skip the reconciliation, since either the control plane migration has been triggered or `gardenlet` cannot reliably determine if the seed is still the owner.
```

Comment:
I'm wondering if it has to be `2m` (or a comparable "small" value). Assuming a seed has ~250 shoots, these timestamp updates will happen ~twice per second. Maybe it has little impact, but still, `Cluster` resources are quite large (embedding `CloudProfile`s, `Shoot`s, etc.), so this will lead to quite some network I/O. This seems especially "painful" under the assumption that such "bad case" migration is very unlikely to happen or to be necessary (at least from today's PoV). I'm not arguing whether we need the "bad case" migration in the first place, no, that's clear, but maybe we should leverage smaller objects like `coordinationv1.Lease`s? Obviously, this comes at the price that extension controllers must read this additional object before the reconciliation (while the `Cluster` is already read anyways).
WDYT?

Reply:
You're right, for the request, network I/O-wise a PATCH is the best and a totally reasonable option. However, the response will always include the entire `Cluster` object, so at least here more data will flow. Also, I don't know which implications it might have on the underlying etcd if such large resources are updated frequently (OTOH, we also have `ShootState`s today, but we don't like the "large nature" of them either). Maybe @timebertt or @amshuman-kr has additional input to this aspect.

> I believe using Cluster should be preferred because it's already read anyways, so having an additional object is both less straightforward (more code), and leads to additional read operations.

Fully agreed, I'd also prefer it for the reasons you mention (given that the "concerns" can be "overcome").

Reply:
The cluster lease mechanism is no longer needed in the latest proposal, so this discussion is no longer relevant.